OpenSim/Simulation/Model/GeometryPath.cpp

```diff
@@ -386,103 +404,55 @@ void GeometryPath::produceForces(const SimTK::State& s,
-            // return NaN for the points and/or throw an Exception- aseth
-            if (dir.norm() < SimTK::SignificantReal){
-                dir = dir*SimTK::NaN;
-            }
-            else{
-                dir = dir.normalize();
-            }
+    // Retains the body index from the previous iteration of the main loop
+    // and the previous direction between the previous point and the current
+    // one (if applicable). This is used to ensure that only one point force
+    // is produced per path point (#3903, #3891).
+    MobilizedBodyIndex previousBodyIndex = MobilizedBodyIndex::Invalid();
+    SimTK::Vec3 previousDirection(0.0);
 
-            force = tension*dir;
-
-            const MovingPathPoint* mppo =
-                dynamic_cast<MovingPathPoint *>(start);
+    const Array<AbstractPathPoint*>& currentPath = getCurrentPath(s);
+    for (int i = 0; i < currentPath.getSize(); ++i) {
 
-            // do the same for the end point of this segment of the path
-            const MovingPathPoint* mppf =
-                dynamic_cast<MovingPathPoint *>(end);
+        const AbstractPathPoint& currentPoint = *currentPath[i];
+        const MobilizedBodyIndex currentBodyIndex = currentPoint.getParentFrame().getMobilizedBodyIndex();
 
-            // add in the tension point forces to body forces
-            if (mppo) {// moving path point location is a function of the state
-                forceConsumer.consumePointForce(s, start->getParentFrame(), mppo->getLocation(s), force);
-            }
-            else {
-                forceConsumer.consumePointForce(s, start->getParentFrame(), start->getLocation(s), force);
-            }
+        SimTK::Vec3 force(0.0);
+        if (previousBodyIndex.isValid() && currentBodyIndex != previousBodyIndex) {
+            force -= tension * previousDirection;
+        }
 
-            if (mppf) {// moving path point location is a function of the state
-                forceConsumer.consumePointForce(s, end->getParentFrame(), mppf->getLocation(s), -force);
+        const AbstractPathPoint* nextPoint = i < currentPath.getSize()-1 ? currentPath[i+1] : nullptr;
+        if (nextPoint && nextPoint->getParentFrame().getMobilizedBodyIndex() != currentBodyIndex) {
+            const SimTK::Vec3 currentDirection = directionBetweenPointsInGroundOrNaNIfCoincident(s, currentPoint, *nextPoint);
+            const SimTK::Vec3 currentToNextForce = tension * currentDirection;
+            force += currentToNextForce;
+
+            // Additionally, account for the work done due to the movement of a `MovingPathPoint`
+            // relative to the body it is on.
+            if (const auto* movingCurrentPoint = dynamic_cast<const MovingPathPoint*>(&currentPoint)) {
+                const Vec3 dPodq_G = currentPoint.getParentFrame().expressVectorInGround(s, currentPoint.getdPointdQ(s));
+                const double fo = ~dPodq_G*currentToNextForce;
```

Comment:
I think that this should apply the total force:
`const double fo = ~dPodq_G*force;`

and you can then remove the lines 438-441
`if (const auto* nextMovingPoint = dynamic_cast<const MovingPathPoint*>(nextPoint)) {... }`

Reply:
Done in [f60706c](https://github.com/opensim-org/opensim-core/pull/3903/commits/f60706c8bf20e8e131de961c64f1dcbd3c1eb8e6)

Reply:
[f60706c](https://github.com/opensim-org/opensim-core/pull/3903/commits/f60706c8bf20e8e131de961c64f1dcbd3c1eb8e6) failed the test suite, so I'm guesing dropping the second `dynamic_cast` is changing behavior.

I've reverted it in 830af43 and, if that passes, I think it's ok to merge what we've got up to now!